packages/sdk/src/protocols/store.ts

```diff
@@ -0,0 +1,317 @@
+      yield messages;
+    }
+  }
+
+  /**
+   * @deprecated Use `queryWithOrderedCallback` instead
+   **/
+  queryOrderedCallback = this.queryWithOrderedCallback;
+
+  /**
+   * Do a query to a Waku Store to retrieve historical/missed messages.
+   *
+   * The callback function takes a `WakuMessage` in input,
+   * messages are processed in order:
+   * - oldest to latest if `options.pageDirection` == { @link PageDirection.FORWARD }
+   * - latest to oldest if `options.pageDirection` == { @link PageDirection.BACKWARD }
+   *
+   * The ordering may affect performance.
+   * The ordering depends on the behavior of the remote store node.
+   * If strong ordering is needed, you may need to handle this at application level
+   * and set your own timestamps too (the WakuMessage timestamps are not certified).
+   *
+   * @throws If not able to reach a Waku Store peer to query,
+   * or if an error is encountered when processing the reply,
+   * or if two decoders with the same content topic are passed.
+   */
+  async queryWithOrderedCallback<T extends IDecodedMessage>(
+    decoders: IDecoder<T>[],
+    callback: (message: T) => Promise<void | boolean> | boolean | void,
+    options?: waku_store.QueryOptions
+  ): Promise<void> {
+    for await (const promises of this.queryGenerator(decoders, options)) {
+      if (await this.processMessages(promises, callback, options)) break;
+    }
+  }
+
+  /**
+   * Do a query to a Waku Store to retrieve historical/missed messages.
+   * The callback function takes a `Promise<WakuMessage>` in input,
+   * useful if messages need to be decrypted and performance matters.
+   *
+   * The order of the messages passed to the callback is as follows:
+   * - within a page, messages are expected to be ordered from oldest to most recent
+   * - pages direction depends on { @link QueryOptions.pageDirection }
+   *
+   * Do note that the resolution of the `Promise<WakuMessage | undefined` may
+   * break the order as it may rely on the browser decryption API, which in turn,
+   * may have a different speed depending on the type of decryption.
+   *
+   * @throws If not able to reach a Waku Store peer to query,
+   * or if an error is encountered when processing the reply,
+   * or if two decoders with the same content topic are passed.
+   */
+  async queryWithPromiseCallback<T extends IDecodedMessage>(
+    decoders: IDecoder<T>[],
+    callback: (
+      message: Promise<T | undefined>
+    ) => Promise<void | boolean> | boolean | void,
+    options?: waku_store.QueryOptions
+  ): Promise<void> {
+    let abort = false;
+    for await (const page of this.queryGenerator(decoders, options)) {
+      const _promises = page.map(async (msgPromise) => {
+        if (abort) return;
+        abort = Boolean(await callback(msgPromise));
+      });
+
+      await Promise.all(_promises);
+      if (abort) break;
+    }
+  }
+
+  createCursor(message: IDecodedMessage): Cursor {
+    if (
```

Comment:
nit: can be simplified to `!message?.timestamp`

Reply:
since we are validating against `timestamp`, `payload` and `contenTopic`, we'd have to do:
```ts
if (!message?.timestamp || !message?.contentTopic || !message?.payload)
```

which imo decreases readability while only reducing one line 

wdyt?

Reply:
to me both seems fine, let's leave it as it is then